srml/authority-discovery/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Authority discovery module.
+//!
+//! This module is used by the `core/authority-discovery` to retrieve the
+//! current set of authorities, learn its own authority id as well as sign and
+//! verify messages to and from other authorities.
+//!
+//! ## Dependencies
+//!
+//! This module depends on the [Im online module](../srml_im_online/index.html)
+//! using its session key.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use app_crypto::RuntimeAppPublic;
+use codec::{Decode, Encode};
+use rstd::prelude::*;
+use srml_support::{decl_module, decl_storage, StorageValue};
+
+pub trait Trait: system::Trait + session::Trait {}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as AuthorityDiscovery {
+		/// The current set of keys that may issue a heartbeat.
+		Keys get(keys): Vec<im_online::AuthorityId>;
+	}
+	add_extra_genesis {
+		config(keys): Vec<im_online::AuthorityId>;
+		build(|
+			  storage: &mut (sr_primitives::StorageOverlay, sr_primitives::ChildrenStorageOverlay),
+			  config: &GenesisConfig
+			  | {
+				  sr_io::with_storage(
+					  storage,
+					  || Module::<T>::initialize_keys(&config.keys),
+				  );
+			  })
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Returns own authority identifier iff it is part of the current authority
+	/// set, otherwise this function returns None. The restriction might be
+	/// softened in the future in case a consumer needs to learn own authority
+	/// identifier in any case.
+	pub fn public_key() -> Option<im_online::AuthorityId> {
+		let authorities = Keys::get();
+
+		let local_keys = im_online::AuthorityId::all();
+
+		let mut intersect: Vec<im_online::AuthorityId> = authorities
+			.into_iter()
+			.filter_map(|authority| {
```

Comment:
Change to find_map, drop collect and the last line.

Reply:
Thanks!